File: app/src/main/java/com/example/myapplication/xstore2/WebSocket.java
package com.example.myapplication.xstore2;


import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.net.InetSocketAddress;
import java.net.Socket;

import javax.net.SocketFactory;
import javax.net.ssl.SSLSocketFactory;

class WebSocket {

    String webSocketEndpoint;
    int webSocketPort;

    public WebSocket(String webSocketEndpoint, int webSocketPort) {
        this.webSocketEndpoint = webSocketEndpoint;
        this.webSocketPort = webSocketPort;

        initSocketWriter();
    }

    protected WebSocket() {
    }

    private Socket socketClient;

    private InetSocketAddress _socketAddress;

    private InetSocketAddress getInetSocketAddress() {
        return _socketAddress;
    }

    private PrintStream _apiSocketWriter;

    private PrintStream getSocketWriter() {
        return _apiSocketWriter;
    }

    private void initSocketWriter() {
        _socketAddress = new InetSocketAddress(webSocketEndpoint, webSocketPort);
        SocketFactory socketFactory = SSLSocketFactory.getDefault();
        InetSocketAddress socketAddress = getInetSocketAddress();
        try {
            socketClient = socketFactory.createSocket(socketAddress.getAddress(), socketAddress.getPort());
        } catch (IOException exception) {
            exception.printStackTrace();
        }
        try {
            _apiSocketWriter = new PrintStream(socketClient.getOutputStream());
        } catch (IOException exception) {
            exception.printStackTrace();
        }
    }

    private BufferedReader _socketReader;

    private BufferedReader getSocketReader() {
        if (_socketReader == null) {
            InputStream streamingReadStream = null;
            try {
                streamingReadStream = socketClient.getInputStream();
            } catch (IOException exception) {
                exception.printStackTrace();
            }
            _socketReader = new BufferedReader(new InputStreamReader(streamingReadStream));
        }
        return _socketReader;
    }

    public void sendMessage(String message) {
        getSocketWriter().print(message);
    }

    public JSONObject getNextMessage() throws JSONException, IOException {
        String line;
        StringBuilder response = new StringBuilder();
        line = getSocketReader().readLine();
        do {
            response.append(line);
            line = getSocketReader().readLine();
        } while (!line.equals(""));
        return new JSONObject(response.toString());
    }

    public boolean isConnected() {
        return !socketClient.isClosed();
    }

    public void disconnect() throws IOException {
        socketClient.close();
    }
}

File: app/src/main/java/com/example/myapplication/xstore2/WebSocket.kt
package com.example.myapplication.xstore2

//import com.example.myapplication.xstore2.model.Movie
import java.io.*
import java.net.InetSocketAddress
import java.net.Socket
import javax.net.ssl.SSLSocketFactory


internal open class WebSocket {
    private var webSocketEndpoint: String? = null
    private var webSocketPort = 0

    constructor(webSocketEndpoint: String?, webSocketPort: Int) {
        this.webSocketEndpoint = webSocketEndpoint
        this.webSocketPort = webSocketPort
        initSocketWriter()
    }

    protected constructor()

    private var socketClient: Socket? = null
    private var socketWriter: PrintStream? = null
    private fun initSocketWriter() {
        val socketAddress = InetSocketAddress(webSocketEndpoint, webSocketPort)
        val socketFactory = SSLSocketFactory.getDefault()
        try {
            socketClient = socketFactory.createSocket(socketAddress.address, socketAddress.port)
        } catch (exception: IOException) {
            exception.printStackTrace()
        }
        socketWriter = PrintStream(socketClient!!.getOutputStream())
    }

    private var _socketReader: BufferedReader? = null
    private val socketReader: BufferedReader
        get() {
            if (_socketReader == null) {
                var streamingReadStream: InputStream? = null
                try {
                    streamingReadStream = socketClient!!.getInputStream()
                } catch (exception: IOException) {
                    exception.printStackTrace()
                }
                _socketReader = BufferedReader(InputStreamReader(streamingReadStream))
            }
            return _socketReader!!
        }

    open fun sendMessage(message: String?) {
        socketWriter!!.print(message)
    }

    @Throws(IOException::class)
    open fun getNextMessage(): String? {
        val response = StringBuilder()

        do {
            // Sometimes server close the connection. socketClient!!.isOutputShutdown checks if the output is closed
            // Without calling isOutputShutdown the "socketReader.readLine()" method blocks thread (without possibility to close)
            // With isOutputShutdown the "socketReader.readLine()" method returns null if socket is closed
            // In both cases isOutputShutdown returns false
            var x1 = socketClient!!.isOutputShutdown
            val line = socketReader.readLine() ?: throw IOException("OutputShutdown Exception - probably webSocket closed by server")
            response.append(line)
        } while (line == "")
        return response.toString()
    }

    open val isConnected
        get() = !socketClient!!.isClosed

    @Throws(IOException::class)
    open fun disconnect() {
        socketClient!!.close()
    }
}